packages/lending/src/adapter.ts

```diff
@@ -0,0 +1,966 @@
+
+    const [flashLoanLoanLogic, flashLoanRepayLogic] = api.protocols.utility.newFlashLoanAggregatorLogicPair(
+      flashLoanAggregatorQuotation.protocolId,
+      flashLoanAggregatorQuotation.loans.toArray()
+    );
+    collateralSwapLogics.push(flashLoanLoanLogic);
+
+    // ---------- swap ----------
+    const swaper = this.findSwaper([wrappedSrcToken, wrappedDestToken]);
+    const swapQuotation = await swaper.quote({
+      input: flashLoanTokenAmount,
+      tokenOut: wrappedDestToken,
+      slippage: defaultSlippage,
+    });
+
+    const swapTokenLogic = swaper.newSwapTokenLogic(swapQuotation);
+    collateralSwapLogics.push(swapTokenLogic);
+
+    // ---------- supply ----------
+    const supplyLogic = await protocol.newSupplyLogic({ input: swapQuotation.output, marketId });
+
+    collateralSwapLogics.push(supplyLogic);
+
+    afterPortfolio.supply(swapQuotation.output.token, swapQuotation.output.amount);
+
+    // compound v3 base token supplied not support collateral swap
+    if (protocolId === 'compoundv3' && supplyLogic.fields.output)
+      throw new Error('Compound V3 does not support the base token for performing collateral swaps');
+
+    // if not compound v3, compound v3 does not need return and add funds in collateral swap flow
+    if (protocolId !== 'compoundv3') {
+      if (!supplyLogic.fields.output) throw new Error('incorrect supply result');
+      // ---------- return funds ----------
+      const returnLogic = api.protocols.utility.newSendTokenLogic({
+        input: supplyLogic.fields.output,
+        recipient: account,
+      });
+      collateralSwapLogics.push(returnLogic);
+
+      // ---------- add funds ----------
+      const addLogic = api.protocols.permit2.newPullTokenLogic({
+        input: new common.TokenAmount(protocol.toProtocolToken(wrappedSrcToken), srcAmount),
+      });
+      collateralSwapLogics.push(addLogic);
+    }
+
+    // ---------- withdraw ----------
+    const withdrawLogic = await protocol.newWithdrawLogic({
+      output: new common.TokenAmount(wrappedSrcToken, srcAmount),
+      marketId,
+    });
+    collateralSwapLogics.push(withdrawLogic);
+
+    afterPortfolio.withdraw(wrappedSrcToken, srcAmount);
+
+    // ---------- flashloan repay ----------
+    collateralSwapLogics.push(flashLoanRepayLogic);
+
+    // ---------- tx related ----------
+    const estimateResult = await api.estimateRouterData(
+      {
+        chainId: this.chainId,
+        account,
+        logics: collateralSwapLogics,
+      },
+      this.permitType
+    );
+
+    const buildRouterTransactionRequest = (
+      args?: Omit<api.RouterData, 'chainId' | 'account' | 'logics'>,
+      apiKey?: string
+    ): Promise<common.TransactionRequest> =>
+      api.buildRouterTransactionRequest(
+        { ...args, chainId: this.chainId, account, logics: collateralSwapLogics },
+        apiKey ? { 'x-api-key': apiKey } : undefined
```

Comment:
Do other buildRouterTransactionRequest need to provide `apiKey`?
Also, where do we pass the `customFees` value?

Reply:
Added `apiKey` to `buildRouterTransactionRequest` for `debtSwap`  `leverageLong` `leverageShort` `deleverage`

`apiKey` is for `debtSwap` `collateralSwap` `leverageLong` `leverageShort` `deleverage`, other features don't need to pass `apiKey`.

we will pass the `customFees` on client side, like this
```typescript
const lendingInfo = await adapter.getLeverageShort(protocolId, marketId, params, account, portfolio);

const transactionRequest = await lendingInfo.buildRouterTransactionRequest(
  {
    permitSig,
    permitData,
    customFees
  },
   apiKey
);
```